src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,187 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Documents;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Implementation.Adornments;
+using Microsoft.CodeAnalysis.Editor.Implementation.EditAndContinue;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
+{
+    internal class InlineDiagnosticsTag : GraphicsTag
+    {
+        public const string TagID = "inline diagnostics - ";
+        public readonly string ErrorType;
+        public readonly InlineDiagnosticsLocations Location;
+
+        private readonly DiagnosticData _diagnostic;
+        private readonly INavigateToLinkService _navigateToLinkService;
+        private readonly IEditorFormatMap _editorFormatMap;
+
+        public InlineDiagnosticsTag(string errorType, DiagnosticData diagnostic, IEditorFormatMap editorFormatMap, InlineDiagnosticsLocations location, INavigateToLinkService navigateToLinkService)
+            : base(editorFormatMap)
+        {
+            ErrorType = errorType;
+            _diagnostic = diagnostic;
+            Location = location;
+            _navigateToLinkService = navigateToLinkService;
+            _editorFormatMap = editorFormatMap;
+        }
+
+        /// <summary>
+        /// Creates a GraphicsResult object which is the error block based on the geometry and formatting set for the item.
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var id = new Run(_diagnostic.Id);
+            Hyperlink? link = null;
+
+            if (_diagnostic.HelpLink is null || _diagnostic.HelpLink.Length is 0)
+            {
+                block.Inlines.Add(id);
+            }
+            else
+            {
+                link = new Hyperlink(id)
+                {
+                    NavigateUri = new Uri(_diagnostic.HelpLink)
+                };
+
+                link.RequestNavigate += HandleRequestNavigate;
+                block.Inlines.Add(link);
+            }
```

Comment:
please just extract a helper that produces the item that will be added to the block.Inlines.  It will make this much cleaner and easier to understand.